src/index.js

```diff
@@ -1,26 +1,136 @@
+
+const HOLO_HASH_AGENT_PREFIX		= Buffer.from(new Uint8Array([0x84, 0x20, 0x24]).buffer);
+const HOLO_HASH_HEADER_PREFIX		= Buffer.from(new Uint8Array([0x84, 0x29, 0x24]).buffer);
+const HOLO_HASH_ENTRY_PREFIX		= Buffer.from(new Uint8Array([0x84, 0x21, 0x24]).buffer);
+const HOLO_HASH_DNA_PREFIX		    = Buffer.from(new Uint8Array([0x84, 0x2d, 0x24]).buffer);
+
+const getHoloHashPrefix = holoHashType => {
+	let holoHashPrefix;
+	switch (holoHashType) {
+		case "agent":
+			holoHashPrefix = HOLO_HASH_AGENT_PREFIX;
+			break;
+		case "header":
+			holoHashPrefix = HOLO_HASH_HEADER_PREFIX;
+			break;
+		case "entry":
+			holoHashPrefix = HOLO_HASH_ENTRY_PREFIX;
+			break;
+		case "dna":
+			holoHashPrefix = HOLO_HASH_DNA_PREFIX;
+			break;
+		default:
+			throw new Error("Received unsupported HoloHash Type in Codec.Digest : ", holoHashType);
+	}
+	return holoHashPrefix;
+}
+
+function check_length (buf, expectedLength) {
+	if (Buffer.byteLength(buf) !== expectedLength)
+		throw new Error(`Unexpected buffer length of ${Buffer.byteLength(buf)}.  Buffer should be ${expectedLength} bytes.`);
+	return buf;
+}
+
+function convert_b64_to_holohash_b64 (rawBase64) {
+	let holoHashbase64 = '';
+	const len = rawBase64.length;
+	for (let i = 0; i < len; i++) {
+		let char = rawBase64[i];
+		if (char === '/') {
+			char = '_'
+		} else if (char === '+') {
+			char = '-'
+		}
+		holoHashbase64 += char;
+	}
+	return holoHashbase64;
+}
+
+// Generate holohash 4 byte (or u32) dht "location" - used for checksum and dht sharding
+function calc_dht_bytes ( data ) {
+	const digest				= blake.blake2b( data, null, 16 );
+	const dht_part			= Buffer.from([digest[0], digest[1], digest[2], digest[3]])
+
+	for (let i of [4, 8, 12]) {
+		dht_part[0] ^= digest[i];
+		dht_part[1] ^= digest[i + 1];
+		dht_part[2] ^= digest[i + 2];
+		dht_part[3] ^= digest[i + 3];
+	}
+
+	return dht_part;
+}
 
 const Codec = {
-    "AgentId": {
-        decode: from_hcs0,
-        encode: to_hcs0,
-    },
-    "Base36": {
-	decode: (str) => base36.decode(str),
-	encode: (buf) => base36.encode(Buffer.from(buf)),
-    },
-    "Signature": {
-        decode: (str) => Buffer.from(str, 'base64'),
-        encode: (buf) => Buffer.from(buf).toString('base64'),
-    },
-    "Digest": {
-        decode: (str) => multihash.decode(multihash.fromB58String(str)).digest,
-        encode: (buf) => multihash.toB58String(multihash.encode(Buffer.from(buf), 'sha2-256')),
-    },
+	"AgentId": {
+		decode: (base64) => Codec.HoloHash.decode(base64),
+		encode: (buf) =>  Codec.HoloHash.encode("agent", Buffer.from(buf)),
+		decodeToHoloHash:(base64) => {
+			const buf = Buffer.from(base64.slice(1), "base64");
+			check_length(Buffer.from(buf), 39);
+			return buf;
+		},
+	},
+	"Base36": {
+		decode: (str) => base36.decode(str),
+		encode: (buf) => base36.encode(Buffer.from(buf)),
+	},
+	"HoloHash": {
+		holoHashStringFromB64: (base64) => convert_b64_to_holohash_b64(base64),
+		holoHashFromBuffer: (holoHashType, buf) => {
```

Comment:
Similar inquiry here. What is the use case for this function? 
Would there be an instance where we would need a holoHash without `u`?

Reply:
This is being used in tests for chaperone and envoy.